### PR TITLE
Adding notification sharding using dynamodb

### PR DESCRIPTION
**Pull request by markrichardsg**

### CODE REVIEW

File: common/src/main/scala/models/Notification.scala [hunk 1]
```diff
@@ -66,6 +68,25 @@ object BreakingNewsNotification {
...
 
+case class NewsstandShardNotification(
+                                       id: UUID,
+                                       shard:Int
+                                     ) extends Notification {
+  override def title = ""
+  override def message: String = ""
+  override def sender: String = "newsstand-shard"
+  override def importance: Importance = Importance.Minor
+  override def withTopics(topics: Set[Topic]): Notification = this
```
Comment (alexduf) [inline common/src/main/scala/models/Notification.scala]: do you mind implementing that? It's short and would avoid spending time searching why it doesn't work if we suddenly decide to use it

File: notification/app/notification/services/azure/NewsstandSender.scala [hunk 2]
```diff
@@ -1,19 +1,40 @@
...
+import scala.concurrent.{ExecutionContext, Future}
 
+
+class NewsstandSender(
+                       hubClient: NotificationHubClient,
+                       legacyNewsstandRegistrationConverterConfig:NewsstandShardConfig,
+                       notificationSchedulePersistence: NotificationSchedulePersistenceAsync,
+                       clock: Clock = Clock.systemUTC()
+                     ) {
+  implicit  val executionContext:ExecutionContext = ExecutionContext.Implicits.global
```
Comment (alexduf) [inline notification/app/notification/services/azure/NewsstandSender.scala]: Given that we're in the context of a play app, I'd rather you get the execution context from play through an implicit parameter at the class level here.

File: notification/app/notification/services/azure/NewsstandSender.scala [hunk 3]
```diff
@@ -1,19 +1,40 @@
...
       tags = Some(Tags.fromTopics(Set(Topic(Newsstand, "newsstand"))))
     )
-    hubClient.sendNotification(push)
+    val shards = legacyNewsstandRegistrationConverterConfig.shards
+    val nowSeconds = Instant.now(clock).getEpochSecond
+
+    hubClient.sendNotification(push).flatMap(hubResult => {
+      Future.sequence{Range(1,shards.toInt + 1).map(shard => notificationSchedulePersistence.writeAsync(NotificationsScheduleEntry(
+        new UUID(id.getMostSignificantBits, id.getLeastSignificantBits+shard).toString,
+        shard.toString,
```
Comment (alexduf) [inline notification/app/notification/services/azure/NewsstandSender.scala]: I think you may be missing the json serialisation of the notification, or have I missed something myself?

File: notification/test/notification/services/azure/NewsstandSenderTest.scala [hunk 4]
```diff
@@ -1,32 +1,87 @@
...
+          current = queue.poll()
+        }
+        val nowSeconds = instant.getEpochSecond
+        val sevenDaysFromNowSeconds = nowSeconds + 604800
+        val uuidMsb = randomId.getMostSignificantBits
+        val uuidLsb = randomId.getLeastSignificantBits
+        notificationsScheduleEntries must be equalTo Set(
+          NotificationsScheduleEntry(
+            new UUID(uuidMsb, uuidLsb + 1).toString,
+            1.toString,
```
Comment (alexduf) [inline notification/test/notification/services/azure/NewsstandSenderTest.scala]: Shouldn't that be the serialised json of the notification?

File: common/src/main/scala/models/Notification.scala (hunk 1 above)
Comment (alexduf) [reply common/src/main/scala/models/Notification.scala]: Nah after thinking about, don't bother

File: notification/app/notification/services/azure/NewsstandSender.scala (hunk 2 above)
Comment (markrichardsg) [reply notification/app/notification/services/azure/NewsstandSender.scala]: Agreed

File: notification/app/notification/services/azure/NewsstandSender.scala (hunk 3 above)
Comment (markrichardsg) [reply notification/app/notification/services/azure/NewsstandSender.scala]: Adding that too

File: notification/test/notification/services/azure/NewsstandSenderTest.scala (hunk 4 above)
Comment (markrichardsg) [reply notification/test/notification/services/azure/NewsstandSenderTest.scala]: Adding that too

File: schedulelambda/riff-raff.yaml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+stacks: [mobile-notifications]
+regions: [eu-west-1]
+deployments:
+  schedule:
+    type: aws-lambda
+    parameters:
+      bucket: mobile-dist
```
Comment (alexduf) [inline schedulelambda/riff-raff.yaml]: `mobile-notifications-dist`

Comment (markrichardsg) [reply schedulelambda/riff-raff.yaml]: changed

**Pull request merged**